### ISSUE TITLE
vault-secrets-webhook: TLS handshake error 

### ISSUE THREAD

**Issue by wiltonfelix** [banzaicloud/bank-vaults#952]
Hello @bonifaido, how are you?

I get that error after 3 weeks using the vault and webhook, 

sets:
k8s: 1.14.10-gke.27 
vault: v1.3.2
webhook: v1.0.0

***Values vault***:
```
replicaCount: 1
strategy:
  type: RollingUpdate
image:
  repository: vault
  tag: 1.3.2
  pullPolicy: IfNotPresent

service:
  name: vault
  type: ClusterIP
  port: 8200

ingress:
  enabled: true
  hosts:
   - xxxxxxx
  annotations:
    kubernetes.io/ingress.class: nginx-private
    kubernetes.io/tls-acme: "true"
    nginx.ingress.kubernetes.io/backend-protocol: "HTTPS"
  tls:
    - secretName: xxxxx
      hosts:
        - xxxxxx


persistence:

  enabled: false

  hostPath: ""

  size: 10G
  accessMode: ReadWriteOnce
podAnnotations:
  prometheus.io/scrape: "true"
  prometheus.io/path: "/metrics"
  prometheus.io/port: "9102"

labels: {}
  #  team: taki

serviceAccount:
  annotations: {}


resources:
  requests:
    cpu: 100m
    memory: 128Mi
  # limits:
  #   cpu: 100m
  #   memory: 128Mi
  # requests:
  #   cpu: 100m
  #   memory: 128Mi

nodeSelector:
  environment: production

tolerations: []

tls:
  secretName: ""

vault:
  customSecrets:
    - secretName: vault-credetials
      mountPath: /etc/gcp
  envSecrets: []
    # - secretName: mysql
    #   secretKey: mysql-username
    #   envName: "MYSQL_USERNAME"
  # Allows adding various custom eenvironment variables directly.
  envs:
    - name: GOOGLE_APPLICATION_CREDENTIALS
      value: /etc/gcp/vault.json
    - name: GOOGLE_PROJECT
      value: vault
    - name: GOOGLE_REGION
      value: us-east1
  externalConfig:
    policies:
      - name: allow_secrets
        rules: path "secret/*" {
                capabilities = ["create", "read", "update", "delete", "list"]
              }
      - name: allow_production
        rules: path "production/*" {
                capabilities = ["create", "read", "update", "delete", "list"]
              }
      - name: allow_staging
        rules: path "staging/*" {
                capabilities = ["create", "read", "update", "delete", "list"]
              }
    auth:
      - type: kubernetes
        roles:
          - name: apps_default_allow
            bound_service_account_names: default
            bound_service_account_namespaces: default
            policies: allow_secrets
            ttl: 1h
          - name: apps_production_allow
            bound_service_account_names: production
            bound_service_account_namespaces: production
            policies: allow_production
            ttl: 1h
          - name: apps_staging_allow
            bound_service_account_names: apps-staging 
            bound_service_account_namespaces: staging
            policies: allow_staging
            ttl: 1h
    secrets:
      - path: secret
        type: kv
        description: General secrets.
        options:
          version: 2
      - path: production
        type: kv
        description: General secrets to production.
        options:
          version: 2
      - path: staging
        type: kv
        description: General secrets to staging.
        options:
          version: 2
  config:
    listener:
      tcp:
        address: '[::]:8200'
        # tls_disable: true
        tls_cert_file: /vault/tls/server.crt
        tls_key_file: /vault/tls/server.key

    telemetry:
      statsd_address: localhost:9125

    ui: true
    storage: 
      # file:
      #   path: "/vault/file"
      gcs:
        bucket: "vault-backend"
        # credentials_file: ""
        ha_enabled: "true"
    api_addr: https://vault:8200

  defaultStorage:
    storage:
      file:
        path: "/vault/file"

  logLevel: info

unsealer:
  image:
    repository: banzaicloud/bank-vaults
    tag: 1.0.0
    pullPolicy: IfNotPresent
  args: [
      "--mode",
      "google-cloud-kms-gcs",
      "--google-cloud-kms-key-ring",
      "vault-auto-unseal",
      "--google-cloud-kms-crypto-key",
      "vault-init",
      "--google-cloud-kms-location",
      "us-east1",
      "--google-cloud-kms-project",
      "vault",
      "--google-cloud-storage-bucket",
      "vault-backend"
  ]
  metrics:
    enabled: true
    debug: true
    name: metrics
    type: ClusterIP
    port: 9091
    serviceMonitor:
      enabled: false
      additionalLabels: {}
    annotations:
      prometheus.io/scrape: "true"
      prometheus.io/path: "/metrics"
      prometheus.io/port: "9091"

statsd:
  image:
    repository: prom/statsd-exporter
    tag: latest
    pullPolicy: IfNotPresent
  metrics:
    enabled: true
    port: 9102
    serviceMonitor:
      enabled: false
      additionalLabels: {}
  config:
    mappings:
    - match: vault.route.*.*
      name: "vault_route"
      labels:
        method: "$1"
        path: "$2"

rbac:
  enabled: true
  psp:
    enabled: false

vaultConfig:
  resources: {}

vaultUnsealer:
  resources:
    requests:
      cpu: 50m
      memory: 64Mi

vaultConfigurer:
  resources:
    requests:
      cpu: 50m
      memory: 64Mi

prometheusStatsdExporter:
  resources:
    requests:
      cpu: 50m
      memory: 128Mi
```
***webhook*** 
```
replicaCount: 3

debug: true

certificate:
    useCertManager: false
    generate: true
    server:
        tls:
            crt:
            key:
    ca:
        crt:
image:
  repository: banzaicloud/vault-secrets-webhook
  tag: 1.0.0
  pullPolicy: IfNotPresent
  imagePullSecrets: []

service:
  name: vault-secrets-webhook
  type: ClusterIP
  externalPort: 443
  internalPort: 8443

env:
  VAULT_IMAGE: vault:1.3.2
  VAULT_ENV_IMAGE: banzaicloud/vault-env:1.0.0
  # VAULT_CAPATH: /vault/tls
  # used when the pod that should get secret injected does not specify
  # VAULT_CLIENT_TIMEOUT: 10s
  DEFAULT_IMAGE_PULL_SECRET: vault-webhook-gcr
  DEFAULT_IMAGE_PULL_SECRET_NAMESPACE: production

metrics:
  enabled: false
  port: 8443
  serviceMonitor:
    enabled: false
    scheme: https
    tlsConfig:
      insecureSkipVerify: true

volumes: []
# - name: vault-tls
#   secret:
#     secretName: vault-tls

volumeMounts: []
# - name: vault-tls
#   mountPath: /vault/tls

podAnnotations: {}

labels: {}
  #  team: taki

resources: {}

tolerations: []

affinity: {}

## Assign a PriorityClassName to pods if set
priorityClassName: ""

rbac:
  enabled: true
  psp:
    enabled: false

# A list of Kubernetes resource types to mutate as well:
# Example: ["ingresses", "servicemonitors"] 
customResourceMutations: []

customResourcesFailurePolicy: Ignore

# This can cause issues when used with Helm, so it is not enabled by default
configMapMutation: false

configmapFailurePolicy: Ignore

podsFailurePolicy: Ignore

secretsFailurePolicy: Ignore

apiSideEffectValue: NoneOnDryRun

namespaceSelector:
  matchExpressions:
  - key: name
    operator: NotIn
    values:
    - kube-system
  # matchLabels:
  #   vault-injection: enabled

# In case of the K8s cluster version is above 1.15 objectSelector is usable
objectSelector: {}
  # matchExpressions:
  # - key: security.banzaicloud.io/mutate
  #   operator: NotIn
  #   values:
  #   - skip

podDisruptionBudget:
  enabled: true
  minAvailable: 1

```
***logs vault pod***
```
020-04-15T15:26:07.637Z [INFO]  core.cluster-listener: starting listener: listener_address=[::]:8201
2020-04-15T15:26:07.637Z [INFO]  core.cluster-listener: serving cluster requests: cluster_listen_address=[::]:8201
2020-04-15T15:26:07.637Z [INFO]  core: vault is unsealed
2020-04-15T15:26:07.637Z [INFO]  core: entering standby mode
2020-04-15T15:26:12.687Z [INFO]  core: acquired lock, enabling active operation
2020-04-15T15:26:13.342Z [INFO]  core: post-unseal setup starting
2020-04-15T15:26:13.446Z [INFO]  core: loaded wrapping token key
2020-04-15T15:26:13.446Z [INFO]  core: successfully setup plugin catalog: plugin-directory=
2020-04-15T15:26:13.602Z [INFO]  core: successfully mounted backend: type=system path=sys/
2020-04-15T15:26:13.603Z [INFO]  core: successfully mounted backend: type=identity path=identity/
2020-04-15T15:26:13.675Z [INFO]  core: successfully mounted backend: type=kv path=secret/
2020-04-15T15:26:13.784Z [INFO]  core: successfully mounted backend: type=kv path=production/
2020-04-15T15:26:13.868Z [INFO]  core: successfully mounted backend: type=kv path=staging/
2020-04-15T15:26:13.868Z [INFO]  core: successfully mounted backend: type=cubbyhole path=cubbyhole/
2020-04-15T15:26:14.679Z [INFO]  core: successfully enabled credential backend: type=token path=token/
2020-04-15T15:26:14.679Z [INFO]  core: successfully enabled credential backend: type=kubernetes path=kubernetes/
2020-04-15T15:26:14.680Z [INFO]  core: successfully enabled credential backend: type=userpass path=userpass/
2020-04-15T15:26:14.680Z [INFO]  core: restoring leases
2020-04-15T15:26:14.680Z [INFO]  rollback: starting rollback manager
2020-04-15T15:26:14.956Z [INFO]  expiration: lease restore complete
2020-04-15T15:26:14.996Z [INFO]  identity: entities restored
2020-04-15T15:26:15.019Z [INFO]  identity: groups restored
2020-04-15T15:26:15.092Z [INFO]  core: post-unseal setup complete
2020-04-15T15:26:16.206Z [INFO]  core: mount tuning of options: path=secret/ options=map[version:2]
2020-04-15T15:26:16.266Z [INFO]  core: mount tuning of options: path=production/ options=map[version:2]
2020-04-15T15:26:16.319Z [INFO]  core: mount tuning of options: path=staging/ options=map[version:2]
2020-04-15T15:28:28.078Z [INFO]  http: TLS handshake error from 10.250.0.120:42148: remote error: tls: bad certificate
2020-04-15T15:28:29.344Z [INFO]  http: TLS handshake error from 10.250.0.120:42152: remote error: tls: bad certificate
2020-04-15T15:28:30.939Z [INFO]  http: TLS handshake error from 10.250.5.20:38862: remote error: tls: bad certificate
2020-04-15T15:28:31.811Z [INFO]  http: TLS handshake error from 10.250.0.120:42154: remote error: tls: bad certificate
2020-04-15T15:28:32.241Z [INFO]  http: TLS handshake error from 10.250.5.20:38864: remote error: tls: bad certificate
2020-04-15T15:28:32.819Z [INFO]  http: TLS handshake error from 10.250.0.120:42158: remote error: tls: bad certificate
2020-04-15T15:28:33.949Z [INFO]  http: TLS handshake error from 10.250.0.120:42160: remote error: tls: bad certificate
2020-04-15T15:28:34.607Z [INFO]  http: TLS handshake error from 10.250.5.20:38874: remote error: tls: bad certificate
2020-04-15T15:28:35.613Z [INFO]  http: TLS handshake error from 10.250.5.20:38878: remote error: tls: bad certificate
2020-04-15T15:28:36.216Z [INFO]  http: TLS handshake error from 10.250.0.120:42164: remote error: tls: bad certificate
2020-04-15T15:28:36.784Z [INFO]  http: TLS handshake error from 10.250.5.20:38880: remote error: tls: bad certificate
2020-04-15T15:28:37.225Z [INFO]  http: TLS handshake error from 10.250.0.120:42166: remote error: tls: bad certificate
2020-04-15T15:28:39.154Z [INFO]  http: TLS handshake error from 10.250.0.120:42172: remote error: tls: bad certificate
2020-04-15T15:28:39.482Z [INFO]  http: TLS handshake error from 10.250.5.20:38884: remote error: tls: bad certificate
```
***webhook***
```
 kubectl logs vault-integration-vault-secrets-webhook-578f655867-bshx4 -n production
time="2020-04-15T15:56:25Z" level=info msg="Failed to request new Vault token Put \"https://vault:8200/v1/auth/kubernetes/login\": x509: certificate signed by unknown authority"
time="2020-04-15T15:56:30Z" level=info msg="Failed to request new Vault token Put \"https://vault:8200/v1/auth/kubernetes/login\": x509: certificate signed by unknown authority"
time="2020-04-15T15:56:32Z" level=fatal msg="failed to create vault clienttimeout [10s] during waiting for Vault token" app=vault-env
```

**Comment by wiltonfelix** [comment]
issue #305 information about that problem.

**Comment by wiltonfelix** [comment]
the ips in the log are from the webhook containers

```
[will@unknown vault-secrets-webhook]$ kubectl get pods --all-namespaces -o wide  |grep 10.250.0.123
production     vault-integration-vault-secrets-webhook-578f655867-bf2b4          0/1     CrashLoopBackOff   6          9m13s   10.250.0.123   gke-taki-pagamentos-gitlab-ci-cd-ce2d52b3-slm4   <none>           <none>
[will@unknown vault-secrets-webhook]$ kubectl get pods --all-namespaces -o wide  |grep 10.250.7.10
production     vault-integration-vault-secrets-webhook-578f655867-vzm4t          0/1     CrashLoopBackOff   6          9m46s   10.250.7.10    gke-taki-pagamentos-gitlab-ci-cd-2b00cc86-136d   <none>           <none>
```

**Comment by balazsbc** [comment]
HI @wiltonfelix , thank you for your continued interest in our Bank-Vaults project.  This issue seems to be a problem in your environment, and not in our project, which is outside the scope of our community support. If you need further help with guaranteed SLA, please consider signing up for one of our support packages, so we can investigate the problem in detail. Please visit https://banzaicloud.com/products/bank-vaults/

**Comment by wiltonfelix** [comment]
Hum, ok.  just to complement that problem, i will add my logs to webhook pod, I am focused on solving this problem

***logs webhook***
```
 kubectl logs vault-integration-vault-secrets-webhook-578f655867-2h2q4 -n vault
time="2020-04-17T16:50:10Z" level=info msg="Failed to request new Vault token Put \"https://vault:8200/v1/auth/kubernetes/login\": x509: certificate signed by unknown authority"
time="2020-04-17T16:50:15Z" level=info msg="Failed to request new Vault token Put \"https://vault:8200/v1/auth/kubernetes/login\": x509: certificate signed by unknown authority"
time="2020-04-17T16:50:16Z" level=fatal msg="failed to create vault clienttimeout [10s] during waiting for Vault token" app=vault-env
```

**Comment by wiltonfelix** [comment]
hello @bonifaido ,

after more test, i get that erro because for me webhook conteiner dont repect label in your namespace.

Im use this command to set label in namespace of the webhook

```
WEBHOOK_NS=${WEBHOOK_NS:-production}
kubectl create namespace "${WEBHOOK_NS}"
kubectl label ns "${WEBHOOK_NS}" name="${WEBHOOK_NS}"

```

***logs webhook conteiner working to me ***

```
time="2020-04-17T22:01:00Z" level=debug msg="reviewing request ec7cab64-80f6-11ea-b054-4201ac100009, named: production/" app=vault-secrets-webhook
time="2020-04-17T22:01:00Z" level=debug msg="Successfully connected to the API" app=vault-secrets-webhook
time="2020-04-17T22:01:00Z" level=debug msg="No pod init containers were mutated" app=vault-secrets-webhook
time="2020-04-17T22:01:00Z" level=info msg="found image banzaicloud/vault-secrets-webhook:1.0.0 in cache"
time="2020-04-17T22:01:00Z" level=debug msg="Successfully mutated pod containers" app=vault-secrets-webhook
time="2020-04-17T22:01:00Z" level=debug msg="Successfully appended pod init containers to spec" app=vault-secrets-webhook
time="2020-04-17T22:01:00Z" level=debug msg="Add generic volumes to podspec" app=vault-secrets-webhook
time="2020-04-17T22:01:00Z" level=debug msg="Successfully appended pod spec volumes" app=vault-secrets-webhook
time="2020-04-17T22:01:00Z" level=debug msg="json patch for request ec7cab64-80f6-11ea-b054-4201ac100009: [{\"op\":\"add\",\"path\":\"/spec/volumes/2\",\"value\":{\"emptyDir\":{\"medium\":\"Memory\"},\"name\":\"vault-env\"}},{\"op\":\"add\",\"path\":\"/spec/initContainers\",\"value\":[{\"command\":[\"sh\",\"-c\",\"cp /usr/local/bin/vault-env /vault/\"],\"image\":\"banzaicloud/vault-env:1.0.0\",\"imagePullPolicy\":\"IfNotPresent\",\"name\":\"copy-vault-env\",\"resources\":{\"limits\":{\"cpu\":\"50m\",\"memory\":\"64Mi\"}},\"securityContext\":{\"allowPrivilegeEscalation\":false},\"volumeMounts\":[{\"mountPath\":\"/vault/\",\"name\":\"vault-env\"}]}]},{\"op\":\"add\",\"path\":\"/spec/containers/0/args\",\"value\":[\"/usr/local/bin/vault-secrets-webhook\"]},{\"op\":\"add\",\"path\":\"/spec/containers/0/command\",\"value\":[\"/vault/vault-env\"]},{\"op\":\"add\",\"path\":\"/spec/containers/0/env/9\",\"value\":{\"name\":\"VAULT_JSON_LOG\",\"value\":\"false\"}},{\"op\":\"add\",\"path\":\"/spec/containers/0/env/9\",\"value\":{\"name\":\"VAULT_ENV_PASSTHROUGH\"}},{\"op\":\"add\",\"path\":\"/spec/containers/0/env/9\",\"value\":{\"name\":\"VAULT_IGNORE_MISSING_SECRETS\",\"value\":\"false\"}},{\"op\":\"add\",\"path\":\"/spec/containers/0/env/9\",\"value\":{\"name\":\"VAULT_ROLE\",\"value\":\"vault-integration-vault-secrets-webhook\"}},{\"op\":\"add\",\"path\":\"/spec/containers/0/env/9\",\"value\":{\"name\":\"VAULT_PATH\",\"value\":\"kubernetes\"}},{\"op\":\"add\",\"path\":\"/spec/containers/0/env/9\",\"value\":{\"name\":\"VAULT_SKIP_VERIFY\",\"value\":\"false\"}},{\"op\":\"add\",\"path\":\"/spec/containers/0/env/9\",\"value\":{\"name\":\"VAULT_ADDR\",\"value\":\"https://vault:8200\"}},{\"op\":\"add\",\"path\":\"/spec/containers/0/volumeMounts/2\",\"value\":{\"mountPath\":\"/vault/\",\"name\":\"vault-env\"}}]" app=vault-secrets-webhook
```
look from app and found image on log .

**Comment by gw0** [comment]
I have a completely different setup, but got the same error as @wiltonfelix (`x509: certificate signed by unknown authority`). Here are my findings in case somebody else needs help at debugging his situation...

The problem is that your pod can not verify the TLS certificate of your Vault cluster. It works if you disable TLS checks on the deployed Helm chart with `podAnnotations: vault.security.banzaicloud.io/vault-skip-verify: "true"`.

After some reading of the fine manual it looks like many images do not contain any trusted Mozilla certificates and `vault-env` does not provide their own. Consequently, you must extract the CA certificate that signed the TLS certificate for your Vault cluster and pass it in (`podAnnotations: vault.security.banzaicloud.io/vault-tls-secret: "vault-ca-cert"`) using a Kubernetes secret under `data: ca.crt`.

Otherwise, I would comment that the `vault-secrets-webhook` is a good piece of software and most confusions come from the lack of a big picture of how all the components work together. Or at least a detailed description of how `vault-secrets-webhook` communicates with other components (not how to use the webhook).

**Comment by cablespaghetti** [comment]
@wiltonfelix I think there might be a few different problems going on here so I hope I'm not further confusing things. Does this problem start when the certificate used by Vault is renewed/changed?

There is an issue which we've noticed which is that when the certificate used by Vault (from a Kubernetes secret) is renewed, Vault does not automatically pick up the new certificate file. 

The symptoms of this issue are that new pods using the webhook/vault-env stop working because they use the new certificate and Vault is still presenting the old one to clients (with an error like yours). If one of the Vault pods gets evicted/replaced it would also have a certificate mismatch from the other pods for the same reason.

There are two ways I've found to fix this at the moment:
1. Add the vault-skip-verify annotation to work around the problem (although the issue between multiple Vault pods might still be an issue)
2. Use https://github.com/stakater/Reloader to automatically restart the Vault pods when the certificate Secret changes

I am going to have a look at the codebase of bank-vaults when I have time to see if the could be built into the vault-configurer or fixed in some other to avoid having to replace the pods.

@bonifaido I'm fairly sure this is the same problem...but can open a separate issue if you prefer.
